ArduPilot/ardupilot
# Custom Controller Support

This PR is part of the GSOC Simulink custom controller implementation project. The aim is to provide a systematic way to run a custom(secondary) controller with very little modification in the original attitude controller and vehicle code.  For this PR only the attitude controller is assumed to be customized. For convenience, original-primary means existing controller and custom-secondary means new controller. 

This PR has the following feature
- flag to compile out the changes in original code.
- In-flight run time transition between original and custom controller with RC switch.
- reset controller to avoid integrator build-up in the offline controller. 
- ground and inflight spool state separation to avoid build-up during arming and takeoff with the custom controller
- bitmask to choose which axis to run the custom controller for
- enable parameter to avoid accidental trigger via RC

I also added an example PID controller where a single PID is used for the attitude control with nonsense initial parameters. In the example, the custom controller gets its already input-shaped target from the AC_AttitudeControl library. I also check which functions in AC_AtttitudeControl or AC_AtttitudeControl_Multi need to be overloaded. Almost all of the functions inside AC_AtttitudeControl library construct target attitude from different reference attitude definitions that are used in various flight modes, ex: input_euler_angle_roll_pitch_yaw, input_rate_bf_roll_pitch_yaw. Since these functions are used to interpret the reference to generate the target attitude, they are not directly part of the controller. reset_yaw_target_and_rate and inertial_frame_reset don't affect the controller because they only change the current attitude/yaw target.

I also find that not running the original controller means that ATC_THR_MIX_MIN, ATC_THR_MIX_MAN, ATC_THR_MIX_MAX have no effect during different stages of flight if we are flying with a custom controller. Because update_throttle_rpy_mix function runs inside rate_controller_run. So, it might be a good idea to run the primary controller regardless or move update_throttle_rpy_mix out of rate_controller_run function.

Pros and cons I find with this approach.

Pros: 
- easily access the attitude controller and all of its functionality
- follows the similar structure as in Attitude Controller
- almost no modification to the AC_AttitudeController code and small modification to the vehicle code

Cons: 
- custom controller param table might reset original controller parameter values(Tridge will better explain)
- Lack of systematic way to store multiple new controllers that might be introduced in the future
- the possibility of making further changes inside the attitude  controller to better support the custom controller

In addition to all that, I could not get the custom controller parameters to appear on the MAVProxy. Which is another problem that I need to look into. For this reason, the default value of CUST_CNT_ENB and CUST_CNT_MSK parameters were changed to enable the custom controller. Setting RC6_OPTION to 6 and simply raising RC6 channel above 1800 will switch to the custom controller. 

How to test:
1. Set CUSTOMCONTROL_ENABLED inside AC_CustomControl.h to 1 
2. Set ATC_CUST_CNT_ENB parameter to 1 
3. Assign RCx_OPTION to 109. Ex: RC6_OPTION 109
4. arm and takeoff
5. Switch RCx to high value 

No difference can be observed as expected while switching . A message will pop on ground control station when switching between controller.
